### PR TITLE
[Element 1.4] Support multiple data files in a single test script

### PR DESCRIPTION
**Pull request by phileba**
resolve https://tricentisflood.atlassian.net/browse/FLO-1900

### CODE REVIEW

**Comment by honglax** [comment]
Hello @phileba, have you implemented the feature that supports `GLOB` yet?

**Comment by phileba** [comment]
> Hello @phileba, have you implemented the feature that supports `GLOB` yet?

not yet @honglax, but the core function is done, so you can list out all files that matched the glob pattern then add to TestData instance

File: packages/core/src/test-data/TestData.ts [hunk 1]
```diff
@@ -100,7 +132,7 @@ export class TestDataSource<T> {
 	 * @param circular Default: true. Pass `false` to disable.
 	 */
 	public circular(circular = true): TestDataSource<T> {
-		this.feeder.circular(circular)
+		Feeder.getInstance().circular(circular, this.loaders[this.loaders.length - 1].loaderName)
```
Comment (HieuLeKMS) [inline packages/core/src/test-data/TestData.ts:135]: This method only affects the last loader, I think that's wrong.The same to `shuffle` and `filter`

File: packages/core/tests/fixtures/data/users.json [hunk 2]
```diff
@@ -1,31 +1,31 @@
 [
 	{
 		"user": 1,
-		"username": "jonny.tester1@loadtest.io"
```
Comment (HieuLeKMS) [inline packages/core/tests/fixtures/data/users.json:4]: Is there a reason why you made these changes? This is test data so I don't know why these changes are necessary

File: packages/core/src/runtime/Test.ts [hunk 3]
```diff
@@ -170,8 +171,28 @@ export default class Test implements ITest {
...
 			debug('Feeding data')
 			testDataRecord = testData.feed()
-			if (testDataRecord === null) {
-				throw new Error('Test data exhausted, consider making it circular?')
+			const invalidName: string[] = []
+			const validTestData = () => {
+				if (testDataRecord === null) return false
+				if (testData.multiple) {
+					for (const key in testDataRecord) {
+						if (testDataRecord[key] === null) {
```
[the remark this answers is not in the document]
Comment (phileba) [reply packages/core/src/runtime/Test.ts:179]: we are checking `null` here because we choose to return null if the pointer is over the length of data lines `const dataRow = lines[item.pointer++] || null`. 
`testDataRecord[key] === undefined` is ok, but null is more strictly because null must be assigned.
`!testDataRecord[key]` will failed if it's value is `0` or `empty string`.

File: packages/core/src/test-data/Feeder.spec.ts [hunk 4]
```diff
@@ -34,7 +34,7 @@ describe('Feeder', () => {
 			username: 'johnny1@loadtest.io',
 			password: 'correcthorsebatterstaple!',
 		})
-		expect(feeder.isComplete).toBe(false)
+		expect(feeder.isComplete).toBe(true)
```
[the remark this answers is not in the document]
Comment (phileba) [reply packages/core/src/test-data/Feeder.spec.ts]: yep, it is in development cause the requirement has changed

File: packages/core/src/test-data/Feeder.ts [hunk 5]
```diff
@@ -3,104 +3,197 @@ import { knuthShuffle } from 'knuth-shuffle'
...
+
+	public append(lines: T[], loaderName = ''): Feeder<T> {
+		const { instanceID } = this
 
 		if (!lines || lines.length === 0) return this
 
-		let newLines = lines.filter((line, index) =>
-			this.filters.every(func => func(line, index, instanceID)),
+		let loader = this.loaderBy(loaderName)[0]
+		if (!loader) loader = this.config(loaderName)
```
[the remark this answers is not in the document]
Comment (phileba) [reply packages/core/src/test-data/Feeder.ts:42]: cool, updated.

File: packages/core/src/test-data/Feeder.ts [hunk 6]
```diff
@@ -3,104 +3,197 @@ import { knuthShuffle } from 'knuth-shuffle'
...
-		}
-
-		this.pointer++
 		return this.peek()
 	}
 
 	/**
 	 * Reads the data at the current cursor without advancing
 	 */
 	public peek(): Option<T> {
```
[the remark this answers is not in the document]
Comment (phileba) [reply packages/core/src/test-data/Feeder.ts:132]: this pointer is designed for a single test data source, if we want to support multiple test data file and multiple test data type combination, we should move this pointer into each test data source instance. And the important thing is we need to make sure the backward compatible.

File: packages/core/src/test-data/TestData.ts [hunk 7]
```diff
@@ -72,26 +68,62 @@ export class TestDataSource<T> {
...
 	 */
 	public setInstanceID(id: string) {
-		this.feeder.instanceID = id
+		Feeder.getInstance().instanceID = id
+	}
+
+	/**
+	 * @internal
+	 * @param name
+	 */
```
[the remark this answers is not in the document]
Comment (phileba) [reply packages/core/src/test-data/TestData.ts:97]: ah I'm going to add some description for this function. will update soon

File: packages/core/src/test-data/TestData.ts (hunk 1 above)
Comment (phileba) [reply packages/core/src/test-data/TestData.ts:151]: yep, it just updates config for each loader, e.g:
```
TestData.fromCSV<Order>('order.csv').circular().as('order_1') => call update config `circular` for loader at index 0, done
TestData.fromCSV<Order>('order-copy.csv').as('order_2').circular() => call update config `circular` for loader at index 1, done
```
then when the test data has loaded, the feeder service will feed data by its config.

File: packages/core/tests/fixtures/data/users.json (hunk 2 above)
Comment (phileba) [reply packages/core/tests/fixtures/data/users.json:4]: when testing multiple files has the same type, we will combine all data together, so as to avoid writing many lines of the test line, I use for loop then the data must be change to make it consistent.
```
const mustFeed = () => ensureDefined(data.feed())
		for (let i = 1; i <= data.size; i++) {
			expect(mustFeed().username).toBe(`samantha${i}@loadtest.io`)
		}
```

File: packages/core/src/test-data/Feeder.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (HieuLeKMS) [reply packages/core/src/test-data/Feeder.ts:132]: I mean this `peek` method reads data where the pointers are, without moving the pointers (as in the method's description). But you made it so that the pointers are updated every time `peek` is called.

**Review by honglax — APPROVED** [review]
